packages/style-engine/class-wp-style-engine.php

```diff
@@ -358,11 +358,11 @@ protected static function is_valid_style_value( $style_value ) {
 		 *
 		 * @return void.
 		 */
-		public static function store_css_rule( $store_name, $css_selector, $css_declarations ) {
+		public static function store_css_rule( $store_name, $css_selector, $css_declarations, $css_at_rule = '' ) {
```

Comment:
I'm not sure I follow. What are the advantages, and what would the API look like in that case?

Reply:
Just catching up — is the question say, what might the output look like if there's more than one rule that belongs to a particular at_rule?

So, theoretically, if we were to have an at rule for a particular query, it'd be good to group them all together into the one at_rule instead of outputting multiple?

Reply:
Sorry, my comment doesn't really make much sense. 😅 I was only thinking out loud and unfiltered of a way to minimize function signature changes.


> what would the API look like in that case?

In terms of what this PR adds to `wp_style_engine_get_stylesheet_from_css_rules` I would change very little, if nothing. That is, passing an `at_rule/container` string.

The class's public methods are also part of the API, but on a lower level and the docs discourage folks from using them directly anyway, but you know... 😄 

>  is the question say, what might the output look like if there's more than one rule that belongs to a particular at_rule?

Yeah, that could be one advantage of handling nested-rule-like objects in a separate class - it could keep track of its child rule, e.g., https://github.com/WordPress/gutenberg/pull/58797/files#diff-f23bfb067de6325fe30b9ec68f807342a3286e40b87937bcb806ea285e0252a3R205

Reply:
> if we were to have an at rule for a particular query, it'd be good to group them all together into the one at_rule instead of outputting multiple

I've been thinking about this, and am not convinced there's much advantage to it. All these styles are generated on the server, and sent to the browser as part of the HTML document. Repetitive strings are well handled by gzip, so there shouldn't be much perf impact.

For these queries in particular, given that they depend on two dynamic values, there's likely to be a fair amount of variation so consolidating won't make a huge difference anyway.

On the other hand, if we did want to nest a bunch of rules under the same @rule, we'd want to make sure that it's always loaded at the very end of all the non-nested rules, because @rules (barring the special case of layers) don't affect specificity. So if we don't want to have to increase the specificity of selectors nested inside them, we need to make sure they always come after non-nested rules.

Reply:
> For these queries in particular, given that they depend on two dynamic values, there's likely to be a fair amount of variation so consolidating won't make a huge difference anyway.

Do you mean the dynamic values in [layout.php](https://github.com/WordPress/gutenberg/pull/58539/files#diff-324697e41855298e2f2c74b078f174e0cbc9075cef736ce9c1e2c169bf64652eR611)?

I'm guessing there could be other potential use cases for this API though.

E.g., `@at-rules` like keyframes or scope with few or no dynamic properties, leading to less variation among @at-rule values.

I know it could be just me 😇 , but I'm also pretty pumped about leaving the door ajar for layers and even other "containers" like nested CSS parent selectors.

Well, maybe not pumped, but mildly curious.

> All these styles are generated on the server, and sent to the browser as part of the HTML document. Repetitive strings are well handled by gzip, so there shouldn't be much perf impact.

Fair point. 

I think that we should still consider batching nested rules if they have the same containing property value in a follow up PR. 

For my part it's more readable and therefore easier for folks to debug. What if there are 10 or 50 repeated instances?

The style engine already groups declarations under the same selector. See the relevant unit [test](https://github.com/WordPress/gutenberg/blob/ef42790c4d60d03435016f6820d36a39a0745ae1/phpunit/style-engine/style-engine-test.php#L672). So the behaviour wouldn't be inconsistent.

> we'd want to make sure that it's always loaded at the very end of all the non-nested rules

Oh yes, great idea. I had that in the back of my mind: processing non-nested rules first then the rest. 

Do you need that feature for this PR? I created a quick test, and I'm not sure it's following that rule 🤔 

<details>

<summary>Example</summary>

```php
public function test_should_return_stylesheet_from_nested_css_rules() {  
    $css_rules = array(  
       array(  
          'selector'     => '.rinceweed',  
          'declarations' => array(  
             'color'        => 'purple',  
          ),  
       ),  
       array(  
          'selector'     => '.saruman',  
          'at_rule'    => '@supports (align-self: stretch)',  
          'declarations' => array(  
             'color'        => 'white',  
             'height'       => '100px',  
             'border-style' => 'solid',  
             'align-self'   => 'stretch',  
          ),  
       ),  
       array(  
          'selector'     => '.saruman',  
          'at_rule'    => '@supports (align-self: stretch)',  
          'declarations' => array(  
             'color'        => 'black',  
             'font-family'  => 'The-Great-Eye',  
          ),  
       ),  
       array(  
          'selector'     => '.gandalf',  
          'at_rule'    => '@supports (border-style: dotted)',  
          'declarations' => array(  
             'color'        => 'grey',  
             'height'       => '90px',  
             'border-style' => 'dotted',  
             'align-self'   => 'safe center',  
          ),  
       ),  
       array(  
          'selector'     => '.radagast',  
          'at_rule'    => '@supports (align-self: stretch)',  
          'declarations' => array(  
             'color'        => 'brown',  
             'height'       => '60px',  
             'border-style' => 'dashed',  
             'align-self'   => 'stretch',  
          ),  
       ),  
       array(  
          'selector'     => '.harry',  
          'declarations' => array(  
             'color'        => 'human',  
          ),  
       ),  
    );  
  
    $compiled_stylesheet = gutenberg_style_engine_get_stylesheet_from_css_rules( $css_rules, array( 'prettify' => false ) );  
  
    $this->assertSame( '.rinceweed{color:purple;}@supports (align-self: stretch){.saruman{color:black;height:100px;border-style:solid;align-self:stretch;font-family:The-Great-Eye;}}@supports (border-style: dotted){.gandalf{color:grey;height:90px;border-style:dotted;align-self:safe center;}}@supports (align-self: stretch){.radagast{color:brown;height:60px;border-style:dashed;align-self:stretch;}}.harry{color:human;}', $compiled_stylesheet );  
}

```

</details>

Reply:
Sure, I'm not saying we shouldn't do it at all, just that there's no particular urgency as it won't make a difference for the current use case.

For layer rules it would be great to have them grouped together. For media queries, I tend to think it's more readable to have rules that apply to the same element grouped, so the current order feels better. But I guess if we have good reason for batching (layers would definitely be that reason) then we should just do it.

> Do you need that feature for this PR? I created a quick test, and I'm not sure it's following that rule

I didn't enforce it in any way, apart from making sure the rules were fed into the style engine in the correct order 😄 
I'm not sure the current implementation would make that easy to enforce? In any case, it seems to work as it is so I think this would be fine for a follow-up.

Reply:
>  just that there's no particular urgency as it won't make a difference for the current use case.

Oh yes, all my babble has been with a quick follow up in mind.

I'm happy to take it on.

It would help in that case to have one or two unit tests to make sure any refactor doesn't break the expected output.

Reply:
...unless we _want_ it to break expected output? In the case of consolidating rules that is. Or would that be an optional setting?

Reply:
> unless we want it to break expected output? In the case of consolidating rules that is.

Oh, I'll break that 😄  

A test that reflects the narrow use case of this PR's layout expectations would be great I think, e.g., building a simple nested rule. Doesn't have to be fancy. Just to have an idea of you want you want to see in the output.

I think consolidation by default would work, but I'll wait for the reviews on the follow up 😄 

Thanks again!

Reply:
Great discussion, folks! The idea of being able to support things like `@keyframes` in the future would be very cool for either a core feature or a plugin that allows folks to build custom animations. For now, though, I quite like the inclusion of the additional `set_at_rule`, `get_at_rule`, and `at_rule` params in this PR — it feels like stuff that likely won't box us in if we were to explore nested at rules and the like further down the track. +1 for adding a couple of quick tests to ensure we've captured the intent for now 👍